pdns/packetcache.cc

```diff
@@ -386,76 +394,79 @@ map<char,int> PacketCache::getCounts()
   return ret;
 }
 
-int PacketCache::size()
-{
-  uint64_t ret=0;
-  for(auto& mc : d_maps) {
-    ReadLock l(&mc.d_mut);
-    ret+=mc.d_map.size();
-  }
-  return ret;
-}
 
-/** readlock for figuring out which iterators to delete, upgrade to writelock when actually cleaning */
 void PacketCache::cleanup()
 {
-  d_statnumentries->store(0);
-  for(auto& mc : d_maps) {
-    ReadLock l(&mc.d_mut);
+  unsigned int maxCached = ::arg().asNum("max-cache-entries");
+  unsigned long cacheSize = *d_statnumentries;
 
-    *d_statnumentries+=mc.d_map.size();
-  }
-  unsigned int maxCached=::arg().asNum("max-cache-entries");
-  unsigned int toTrim=0;
-  
-  unsigned long cacheSize=*d_statnumentries;
-  
+  // two modes - if toTrim is 0, just look through 10%  of the cache and nuke everything that is expired
+  // otherwise, scan first 5*toTrim records, and stop once we've nuked enough
+  unsigned int toTrim = 0, lookAt = 0;
   if(maxCached && cacheSize > maxCached) {
     toTrim = cacheSize - maxCached;
+    lookAt = 5 * toTrim;
+  } else {
+    lookAt = cacheSize / 10;
   }
 
-  unsigned int lookAt=0;
-  // two modes - if toTrim is 0, just look through 10%  of the cache and nuke everything that is expired
-  // otherwise, scan first 5*toTrim records, and stop once we've nuked enough
-  if(toTrim)
-    lookAt=5*toTrim;
-  else
-    lookAt=cacheSize/10;
+  DLOG(L<<"Starting cache clean, cacheSize: "<<cacheSize<<", lookAt: "<<lookAt<<", toTrim: "<<toTrim<<endl);
 
-  //  cerr<<"cacheSize: "<<cacheSize<<", lookAt: "<<lookAt<<", toTrim: "<<toTrim<<endl;
-  time_t now=time(0);
-  DLOG(L<<"Starting cache clean"<<endl);
-  //unsigned int totErased=0;
+  time_t now = time(0);
+  unsigned int totErased = 0;
   for(auto& mc : d_maps) {
     WriteLock wl(&mc.d_mut);
-    typedef cmap_t::nth_index<1>::type sequence_t;
-    sequence_t& sidx=mc.d_map.get<1>();
-    unsigned int erased=0, lookedAt=0;
-    for(sequence_t::iterator i=sidx.begin(); i != sidx.end(); lookedAt++) {
+    auto& sidx = boost::multi_index::get<SequenceTag>(mc.d_map);
+    unsigned int erased = 0, lookedAt = 0;
+    for(auto i = sidx.begin(); i != sidx.end(); lookedAt++) {
       if(i->ttd < now) {
-	sidx.erase(i++);
-	erased++;
-      }
-      else {
-	++i;
+        sidx.erase(i);
+        erased++;
       }
 
       if(toTrim && erased > toTrim / d_maps.size())
-	break;
-      
+        break;
+
       if(lookedAt > lookAt / d_maps.size())
-	break;
+        break;
+
+      i++;
```

Comment:
This looks to be wrong - it refers to i after erasing it. This is why it used to be blah.erase(i++) which moves to the next iterator before erasing it. There is also a new syntax I think i=blah.erase(i) but haven't checked.